Refactor ProcessGroupNCCL collective primitives

mxw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19631 Fix CUDA stream syncing bug in allgather and reduce_scatter
* #18844 Add support for reduce-scatter in c10d
* **#18820 Refactor ProcessGroupNCCL collective primitives**

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

Differential Revision: [D14762369](https://our.internmc.facebook.com/intern/diff/D14762369)

mxw committed:
Refactor ProcessGroupNCCL collective primitives

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
-  // To make sure each tensor is on separate devices
-  std::unordered_set<int> usedDevices;
-  usedDevices.reserve(input.size());
+  auto const& first = tensors.front();
```
Prefer `const auto` over `auto const`

```
$ git grep 'const auto' | wc -l
3782
$ git grep 'auto const' | wc -l
16
```

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
     throw std::runtime_error(
-        "The number of input tensors is larger than "
-        "the number of available GPUs");
+      "Tensor list mustn't be larger than the number of available GPUs"
```
`clang-format` will reformat this with bigger indent

mxw replied:
Does PyTorch use clang-format as a source-of-truth?  I'm freaked out about what it does to lambdas.

mxw replied:
This file is decidedly not clang-formatted, and running clang-format indeed turns the lambda formatting into spaghetti.  I'll leave it for now and defer to the code formatting zealots who surely will visit upon this file if they find it necessary at some point 😬

pietern replied:
There are a few other clang-format violations in c10d right now, so we can address this in a separate commit. Ideally I'd like to setup a clang-format linter for this directory (and others where applicable) so we don't have to deal with partially formatted files.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
-      throw std::runtime_error(
-          "Expecting all input tensors to have identical "
-          "number of elements");
+    if (t.numel() != first.numel()) {
```
Comparing `sizes` is even better, seeing as mixing different dimensions that happen to have the same number of elements is not a good idea.

pietern reviewed (commented):
Nice work, this dedups quite a bit. There are some CI failures to check out though.

mxw committed:
Update on "Refactor ProcessGroupNCCL collective primitives"

Refactor ProcessGroupNCCL collective primitives

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

gh-metadata: pytorch pytorch 18820 gh/mxw/1/head

pietern commented:
@pytorchbot retest this please

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
+/*
+ * Check that all `tensors' have the same type and shape and are distributed
+ * across distinct GPUs.
+ */
```
All comments in this directory use `//` style.

Did a quick check in the rest of the repository and the block comment is not prevalent elsewhere either:

```
$ grep -ir '^//' * | wc -l
3179
$ grep -ir '^/\*' * | wc -l
33
```

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
+/*
+ * Flatten each list in `tensor_lists' for a gather or scatter operation, and
+ * ensure compatibility with the corresponding tensor in `other'.
+ */
```
Comment style.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
+  std::vector<at::Tensor> flattened;
+  flattened.resize(num_devices);
+
+  for (auto i = size_t{}; i < num_devices; ++i) {
```
I've never seen this before and have always used `size_t i`. TIL.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -384,235 +401,164 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
+    // prevent being freed before the collective finishes.
+    //
+    // We only record `inputs' here, and leave recording `outputs' to `fn' for
+    // operations where `inputs' and `outputs' are the same.
```
"[...] where inputs and outputs are **NOT** the same.", otherwise it would be redundant, no?

mxw replied:
Yep—will fix.

pietern reviewed (commented):
Some additional nits.

In general this looks good to me. Thanks for reducing the amount of copy/paste here!

I didn't fully vet the logic in every path and trust that our tests provide with enough coverage.

@mrshenli Can you also take a look?

mxw committed:
Update on "Refactor ProcessGroupNCCL collective primitives"

Refactor ProcessGroupNCCL collective primitives

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

gh-metadata: pytorch pytorch 18820 gh/mxw/1/head

mxw commented:
Most recent update addresses all of @pietern's comments.

mxw commented:
@pytorchbot retest

pietern commented:
@pytorchbot retest this please

mrshenli commented:
@pytorchbot retest this please

pietern commented:
@pytorchbot retest this please

mxw commented:
@pytorchbot retest this please

mrshenli commented:
@mxw you might need a rebase

mxw committed:
Update on "Refactor ProcessGroupNCCL collective primitives"

Refactor ProcessGroupNCCL collective primitives

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

gh-metadata: pytorch pytorch 18820 gh/mxw/1/head

mxw commented:
@pytorchbot retest this please

mrshenli reviewed (approved):
I think the lint failure is irrelevant.

```
57.63s$ flake8
./test/test_fake_quant.py:28:1: E302 expected 2 blank lines, found 1
The command "flake8" exited with 1.
```

@pietern do you have any comments to add?

mxw committed:
Update on "Refactor ProcessGroupNCCL collective primitives"

Refactor ProcessGroupNCCL collective primitives

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

gh-metadata: pytorch pytorch 18820 gh/mxw/1/head
